### ISSUE TITLE
Unable to download Podcasts with airsonic-advanced

### ISSUE THREAD

**Issue by hamersaw** [memen45/SubMusic#62]
I'm trying to set this up on my Forerunner 955 Solar using the [airsonic-advanced docker image](https://hub.docker.com/r/airsonicadvanced/airsonic-advanced#!). The command I'm using to start my airsonic-advanced docker container is listed below: 

```
sudo docker run -d \
    --env JAVA_OPTS=-Dserver.forward-headers-strategy=native \
    --name airsonic \
    --publish 80:4040 \
    --publish 443:4041 \
    --restart unless-stopped \
    --volume /data/@media/music:/var/music \
    --volume /data/@media/podcasts:/var/podcasts \
    --volume /data/@containers/airsonic/playlists:/var/playlists \
    --volume /data/@containers/airsonic/data:/var/airsonic \
    airsonicadvanced/airsonic-advanced:latest
```

Currently, I have the server address in submusic set as `http://192.168.10.200`. The "Test Server" function returns:

`GarminSdkError::-400 Error HTTP --> GarminSdkError::INVALID_HTTP_BODY_IN_NETWORK_RESPONSE`

However, I am periodically (seems to work sometimes and doesn't work others) able to `Sync` playlists, but even when I set podcasts to offline storage they never download any episodes even though the podcasts are listed from the main page and have a `0 -needs sync` decoration. 

From reading other issues it seems like this may be an issue with not exposing HTTPS, is there any way to not run over HTTPS? I'm just exposing this to my internal network and do not want the hassle of setting up certs and additionally it looks like setting up HTTPS on the docker image is complicated. I would be happy to help contribute this with a little guidance.

Is there anything else that could be the issue here? Very hopeful to get this working.

**Comment by hamersaw** [comment]
OK, to hopefully make this easier on the next person. Got everything setup using a reverse proxy with lighttpd and letsencrypt certs. I was able to use the airsonic-advanced docker image using the below cli command:
```
sudo docker run -d \
    --env JAVA_OPTS="-Dserver.servlet.context-path=/airsonic" \
    --name airsonic \
    --publish 4040:4040 \
    --publish 4041:4041 \
    --restart unless-stopped \
    --volume <music-path>:/var/music \
    --volume <podcasts-path>:/var/podcasts \
    --volume <playlists-path>:/var/playlists \
    --volume <config-path>:/var/airsonic \
    airsonicadvanced/airsonic-advanced:latest
```
Most important for my setup is java option `-Dserver.servlet.context-path=/airsonic`. This means that airsonic-advanced will prefix all of it's URL paths with the specified path. The I used added a reverse proxy to the `$SERVER["socket"] == ":443" ` block in my lighttpd configuration using:
```
    $HTTP["url"] =~ "^/airsonic" {
        proxy.server = ( "" => ( "" => ( "host" => "192.168.10.200", "port" => 4040)))
    }
```
So now I can access my airsonic-advanced instance on `https://<domain>/airsonic`.

**Comment by hamersaw** [comment]
@memen45 now I'm running into an issue with syncing podcasts. When I specify enable one to maintain an offload instance and then attempt to sync it downloads a song (rather than a podcast) from my library instead. I captured the HTTP requests from lighttpd and I see this:
```
192.168.10.1 <domain> - [08/Nov/2022:16:45:20 +0000] "GET /airsonic/rest/getPodcasts?v=1.10.2&c=SubMusic-v0.2.6-kynortas&p=<password>&f=json&u=<username>&id=3&includeEpisodes=false HTTP/1.1" 200 1185 "-" "Garmin Forerunner 955 Solar/12.27"
192.168.10.1 <domain> - [08/Nov/2022:16:45:20 +0000] "GET /airsonic/rest/getPodcasts?v=1.10.2&c=SubMusic-v0.2.6-kynortas&p=<password>&f=json&u=<username>id=3 HTTP/1.1" 200 3964 "-" "Garmin Forerunner 955 Solar/12.27"
192.168.10.1 <domain> - [08/Nov/2022:16:45:30 +0000] "GET /airsonic/rest/stream?v=1.10.2&c=SubMusic-v0.2.6-kynortas&p=>password>&f=json&u=<username>&id=584&format=raw HTTP/1.1" 200 6694225 "-" "Garmin Forerunner 955 Solar/12.27"
```
This looks like it issues a request to ensure the podcast channel exists, then to retrieve the podcast episode list, and finally to download the podcast. I think the issue is that, according to the subsonic API definition, the [stream] endpoint uses an ID that is "Obtained by calls to getMusicDirectory.". This would explain why we download a song rather than a podcast. It looks like the [downloadPodcastEpisode](http://www.subsonic.org/pages/api.jsp#downloadPodcastEpisode) endpoint would be more appropriate here.

Does this sound possible? I would be open to contributing a fix on this with a little guidance, do you want me to create a separate issue for this? Also would be VERY interested in an option to specify "keep last n podcasts" as well, rather than just retrieving the latest episode it would be nice to keep the last few.

**Comment by memen45** [comment]
Great to hear you got the connection working! 

The `downloadPodcastEpisode` endpoint is a bit misleading, but it actually returns nothing but an empty response. It is an instruction for the server to fetch the latest podcast file and download it to the server. 

The `stream` endpoint should work for both songs and podcasts. Podcasts should not have the same id as a song. Could you verify the response for the second `getPodcasts` request? Does it have `id=584` as the latest podcast?

**Comment by hamersaw** [comment]
@memen45 that makes sense! I've done some deeper diving on the `id` issue, looking at the airsonic storage DB, HTTP requests / responses, etc. I see that podcasts have multiple ID fields, namely `id`, `channel_id`, and `media_file_id`. I checked to make sure this is [in the original airsonic ](https://github.com/airsonic/airsonic/blob/5ccca059d5cfe3dd19734c27861b434ea21b43d8/airsonic-main/src/main/java/org/airsonic/player/domain/PodcastEpisode.java#L32-L34)and not just an airsonic-advanced construct as well.

I think that the `id` field is specific to the podcast channel (denoted by `channel_id`). In the DB I see that that all of the podcast episodes are available with incremental IDs, even those that have not yet been downloaded. Then those that have been downloaded also have the `media_file_id` field populated. In my example above, the podcast episode ID is 584, but the media file ID is 10. Then when I look at the media files there is a corresponding entry with the podcasts media file ID, the entry using the podcast ID matches the song that is being incorrectly downloaded. When the podcast episode has not been downloaded the media file ID is empty.

IIUC when we retrieve the episode we use the `id` field in the podcast. I believe that is set [here](https://github.com/memen45/SubMusic/blob/bd76a1907933c745d439753af6bc535a272c69fe/source/Api/SubsonicProvider.mc#L299). Does it make sense that this should instead be the `media_file_id` so it correctly maps to the podcasts downloaded media file?

**Comment by memen45** [comment]
Looking at the subsonic api specification, there is no field `media_file_id`. [See the xml under 'Example'.](http://www.subsonic.org/pages/api.jsp#getPodcasts) However, you are correct in that the `id` field seems to be some useless internal id and not the actual id used for streaming. There is a field called `streamId`, which I now think might be the correct field. Could you confirm with your instance? ([Source](https://github.com/airsonic-advanced/airsonic-advanced/blob/89e6fe9aec80daf8224c6696f8f86bfcf22c6e0a/subsonic-rest-api/src/main/resources/subsonic-rest-api.xsd#L435) for airsonic-advanced implementation)

Normally, I test against the Nextcloud implementation of the Subsonic API, and there the `id` actually works on the stream endpoint. I checked again, and the `id` field is actually equal to the `streamId`, hence it worked before. 

So, in short, I think it is best to change `id` to `streamId`.

**Comment by hamersaw** [comment]
@memen45 I bet you are right! I was just looking at the DB, the API probably uses the `streamId` field you mention.

**Comment by memen45** [comment]
Okay thanks, I have changed the code, but I am unable to test it due to a bug in the simulator. Unfortunately, this will delay the update until Garmin fixes it (opened a bug report [here](https://forums.garmin.com/developer/connect-iq/i/bug-reports/bug-makewebrequest-returns-unknown_error-when-in-simulator-media-mode-sync)).

**Comment by memen45** [comment]
I have circumvented the bug for now. You can find the fix in the latest version on Connect IQ (0.2.7-perieres). Cheers!
